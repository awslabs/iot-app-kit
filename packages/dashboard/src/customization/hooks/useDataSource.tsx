import { SiteWiseQuery } from '@iot-app-kit/source-iotsitewise';
import React, { useContext, useState } from 'react';

type DataSource = { name: string; query: SiteWiseQuery | undefined };

type Dispatcher = React.Dispatch<
  React.SetStateAction<{
    name: string;
    query: SiteWiseQuery | undefined;
  }>
>;

/**
 * Context to access a data source provided to the dashboard
 */
const DataSourceContext = React.createContext<{ dataSource: DataSource; update: Dispatcher }>({
  dataSource: { name: 'iotsitewise', query: undefined },
  update: () => {},
});
export const DataSourceProvider: React.FC<{ query?: SiteWiseQuery }> = ({ children, query }) => {
  const [dataSource, setDataSource] = useState({ name: 'iotsitewise', query });

  return (
    <DataSourceContext.Provider value={{ dataSource, update: setDataSource }}>{children}</DataSourceContext.Provider>
  );
};

export const useDataSource = () => {
  const { dataSource, update } = useContext(DataSourceContext);
  return { dataSource, update };
};
