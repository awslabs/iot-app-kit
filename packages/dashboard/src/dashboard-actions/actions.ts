import { Action } from 'redux';
import { Position, DashboardConfiguration, Anchor } from '../types';

export const MOVE = 'MOVE';

export interface MoveAction extends Action<'MOVE'> {
  type: typeof MOVE;
  payload: {
    position: Position;
    prevPosition: Position | undefined;
    widgetIds: string[];
    cellSize: number;
  };
}
export const onMoveAction = (payload: MoveAction['payload']): MoveAction => ({
  type: MOVE,
  payload,
});

export const RESIZE = 'RESIZE';

export interface ResizeAction extends Action<'RESIZE'> {
  type: typeof RESIZE;
  payload: {
    anchor: Anchor;
    changeInPosition: Position;
    widgetIds: string[];
    cellSize: number;
  };
}
export const onResizeAction = (payload: ResizeAction['payload']): ResizeAction => ({
  type: RESIZE,
  payload,
});

export type DashboardAction = MoveAction | ResizeAction;
