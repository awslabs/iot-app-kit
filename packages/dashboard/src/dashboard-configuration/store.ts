import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { nanoid } from 'nanoid';
import max from 'lodash-es/max';
import unionBy from 'lodash-es/unionBy';
import maxBy from 'lodash-es/maxBy';
import minBy from 'lodash-es/minBy';
import xorBy from 'lodash-es/xorBy';
import { v4 } from 'uuid';
import { WidgetPropertiesGeneratorMap } from '#customization/widgetPropertiesGeneratorMap';
import type { RefreshRate } from '#components/refreshRate/types';
import { placeWithinRectangle } from '#grid/rectangle/nest';
import { resizeSelectionBox } from '~/features/widgetResize/resizeSelectionBox';
import { transformWidget } from '~/features/widgetResize/transformWidget';
import { getSelectionBox } from '~/features/widgetSelection/getSelectionBox';
import { moveSelectionBox } from '~/features/widgetSelection/moveSelectionBox';
import { compact } from '@iot-app-kit/helpers';
import type {
  DashboardConfiguration,
  DashboardDisplaySettings,
  DashboardWidget,
} from '#types';
import { createRectangle } from '#grid/rectangle/create';
import type { Anchor } from '#grid/rectangle/types';
import { placeWithinGrid } from '#grid/contain';
import { createGrid } from '#grid/create';
import type { Grid } from '#grid/types';
import { evolve } from '@iot-app-kit/helpers/lists/evolve';
import Widgets from '#components/widgets/list';
import { map } from '@iot-app-kit/helpers/lists/map';
import { LastArrayElement, RequiredKeysOf } from 'type-fest';
import type { Position } from '#grid/position/types';
import { DashboardWidgetType } from '#migration/types';
import invariant from 'tiny-invariant';
import type { F } from 'ts-toolbelt';

declare const pipe: F.Pipe;

export interface DashboardState {
  copiedWidgetIds: string[];
  pasteCounter: number;
  dashboardConfiguration: DashboardConfiguration;
  isCanvasEnabled: boolean;
}

export const initialState: DashboardState = {
  copiedWidgetIds: [],
  pasteCounter: 0,
  dashboardConfiguration: {
    widgets: [],
    displaySettings: {
      numColumns: 100,
      numRows: 100,
      cellSize: 20,
      significantDigits: 4,
    },
    querySettings: {
      refreshRate: 5000,
    },
  },
  isCanvasEnabled: true,
};

export const {
  reducer,
  actions: {
    createWidgets,
    deleteWidgets,
    bringWidgetsToFront,
    sendWidgetsToBack,
    moveWidgets,
    pasteWidgets,
    copyWidgets,
    resizeWidgets,
    updateCellSize,
    updateDecimalPlaces,
    updateDefaultViewport,
    updateRefreshRate,
    updateWidgets,
    updateHeight,
    updateWidth,
    disableCanvas,
    enableCanvas,
    replaceDashboardConfiguration,
  },
} = createSlice({
  name: 'dashboardConfiguration',
  initialState,

  reducers: {
    updateHeight: (state, action: PayloadAction<number>) => {
      state.dashboardConfiguration.displaySettings.numRows = action.payload;
    },

    updateWidth: (state, action: PayloadAction<number>) => {
      state.dashboardConfiguration.displaySettings.numColumns = action.payload;
    },

    updateCellSize: (state, action: PayloadAction<number>) => {
      state.dashboardConfiguration.displaySettings.cellSize = action.payload;
    },

    updateDecimalPlaces: (state, action: PayloadAction<number>) => {
      state.dashboardConfiguration.displaySettings.significantDigits =
        action.payload;
    },

    updateDefaultViewport: (
      state,
      action: PayloadAction<DashboardConfiguration['defaultViewport']>
    ) => {
      state.dashboardConfiguration.defaultViewport = action.payload;
    },

    updateRefreshRate: (state, action: PayloadAction<RefreshRate>) => {
      if (!state.dashboardConfiguration.querySettings) {
        state.dashboardConfiguration.querySettings = {
          refreshRate: action.payload,
        };
      } else {
        state.dashboardConfiguration.querySettings.refreshRate = action.payload;
      }
    },

    createWidgets: (
      state,
      {
        payload,
      }: PayloadAction<{
        widgets: (Position & { type: DashboardWidgetType })[];
      }>
    ) => {
      const {
        cellSize = 20,
        numRows,
        numColumns,
      } = state.dashboardConfiguration.displaySettings;

      const newWidgetsNew = pipe(
        map((widget: Position & { type: DashboardWidgetType }) => ({
          ...widget,
          id: nanoid(),
        })),
        map((widget) => {
          const { properties, initialSize } =
            WidgetPropertiesGeneratorMap[widget.type];

          return {
            ...widget,
            ...initialSize,
            properties,
          };
        })
      )(payload.widgets) satisfies DashboardWidget[];

      const newWidgets = payload.widgets
        .map(({ type, x = 0, y = 0 }) => {
          const { properties, initialSize } =
            WidgetPropertiesGeneratorMap[type];

          const { width: widgetPixelWidth, height: widgetPixelHeight } =
            initialSize || { height: 150, width: 150 };

          return {
            id: nanoid(),
            type,
            width: Math.min(Math.ceil(widgetPixelWidth / cellSize), numColumns),
            height: Math.min(Math.ceil(widgetPixelHeight / cellSize), numRows),
            x,
            y,
            z: 0,
            properties: (properties && properties()) || {},
          };
        })
        .map((widget) => {
          const { position: position, dimensions } = placeWithinRectangle(
            {
              position: { x: widget.x, y: widget.y },
              dimensions: { height: widget.height, width: widget.width },
            },
            {
              position: { x: 0, y: 0 },
              dimensions: { width: numColumns, height: numRows },
            }
          );

          return {
            ...widget,
            ...position,
            ...dimensions,
          };
        });

      state.dashboardConfiguration.widgets.push(...newWidgets);
    },

    updateWidgets: (
      state,
      { payload }: PayloadAction<{ widgets: DashboardWidget[] }>
    ) => {
      const positionWidgets = evolve((widget: DashboardWidget) => {
        return placeWithinGrid(
          widget,
          gridFromDisplaySettings(state.dashboardConfiguration.displaySettings)
        );
      });

      const positionedWidgets = positionWidgets(payload.widgets);

      state.dashboardConfiguration.widgets = unionBy(
        positionedWidgets,
        state.dashboardConfiguration.widgets,
        ({ id }) => id
      );
    },

    deleteWidgets: (state, action: PayloadAction<{ widgetIds: string[] }>) => {
      const remainingWidgets = state.dashboardConfiguration.widgets.filter(
        ({ id }) => !action.payload.widgetIds.includes(id)
      );

      state.dashboardConfiguration.widgets = remainingWidgets;
    },

    copyWidgets: (
      state,
      { payload: { widgetIds } }: PayloadAction<{ widgetIds: string[] }>
    ) => {
      state.copiedWidgetIds = widgetIds;
    },

    pasteWidgets: (
      state,
      {
        payload: { position, widgetIds },
      }: PayloadAction<{ position?: Position | undefined; widgetIds: string[] }>
    ) => {
      const cellSize =
        state.dashboardConfiguration.displaySettings.cellSize ?? 20;
      const copyGroup = getWidgetsFromIds({
        widgetIds,
        widgets: state.dashboardConfiguration.widgets,
      });
      const gridWidth = state.dashboardConfiguration.displaySettings.numColumns;
      const gridHeight = state.dashboardConfiguration.displaySettings.numRows;

      let offset: Position = {
        x: 0,
        y: 0,
      };
      const correctionOffset: Position = {
        x: 0,
        y: 0,
      };
      if (position !== undefined) {
        state.pasteCounter = 0;

        const cellPosition: Position = {
          x: position && Math.floor(position.x / cellSize),
          y: position && Math.floor(position.y / cellSize),
        };

        // getting widgets group's left most cell value
        const leftmostWidget: DashboardWidget =
          minBy(copyGroup, 'x') || copyGroup[0];
        const groupLeftX = leftmostWidget.x;

        // getting widgets group's top most cell value
        const topmostWidget: DashboardWidget =
          minBy(copyGroup, 'y') || copyGroup[0];
        const groupTopY = topmostWidget.y;

        // getting widgets group's right most cell value
        const widgetsRightPos = copyGroup.map((widget) => {
          return widget.x + widget.width;
        });
        const groupRightX = max(widgetsRightPos) || gridWidth;

        // getting widgets group's bottom most cell value
        const widgetsBottomPos = copyGroup.map((widget) => {
          return widget.y + widget.height;
        });
        const groupBottomY = max(widgetsBottomPos) || gridHeight;

        // calculating widgets group's width & height
        const groupWidth = groupRightX - groupLeftX;
        const groupHeight = groupBottomY - groupTopY;

        // setting offset postion
        offset = {
          x: cellPosition.x - groupLeftX,
          y: cellPosition.y - groupTopY,
        };

        // setting correction offset position if widgets group's width or height is going off the grid
        if (cellPosition.x + groupWidth > gridWidth) {
          correctionOffset.x = cellPosition.x + groupWidth - gridWidth;
        }
        if (cellPosition.y + groupHeight > gridHeight) {
          correctionOffset.y = cellPosition.y + groupHeight - gridHeight;
        }
      }

      const widgetsToPaste = copyGroup.map((widget) => ({
        ...widget,
        id: v4(),
        x: offset.x + widget.x + state.pasteCounter - correctionOffset.x,
        y: offset.y + widget.y + state.pasteCounter - correctionOffset.y,
      }));

      state.dashboardConfiguration.widgets.push(...widgetsToPaste);
      state.pasteCounter = position ? 0 : state.pasteCounter;
      // state.selectedWidgetIds = widgetsToPaste.map(({ id }) => id);
    },

    bringWidgetsToFront: (
      state,
      { payload: { widgetIds } }: PayloadAction<{ widgetIds: string[] }>
    ) => {
      // We don't need to do anything if all widgets are selected
      if (widgetIds.length === state.dashboardConfiguration.widgets.length) {
        return state;
      }

      const widgetsToUpdate = getWidgetsFromIds({
        widgetIds,
        widgets: state.dashboardConfiguration.widgets,
      });

      const unselectedWidgets = xorBy(
        state.dashboardConfiguration.widgets,
        widgetsToUpdate,
        'id'
      );

      const topZIndex = maxBy(unselectedWidgets, 'z')?.z ?? 0;
      const minSelectedZ = minBy(widgetsToUpdate, 'z')?.z ?? 0;

      const zOffset = topZIndex + 1 - minSelectedZ;

      const translatedWidgets = state.dashboardConfiguration.widgets.map(
        (widget) => ({
          ...widget,
          z: widgetIds.includes(widget.id) ? widget.z + zOffset : widget.z,
        })
      );

      state.dashboardConfiguration.widgets = translatedWidgets;
    },

    sendWidgetsToBack: (
      state,
      { payload: { widgetIds } }: PayloadAction<{ widgetIds: string[] }>
    ) => {
      // We don't need to do anything if all widgets are selected
      if (widgetIds.length === state.dashboardConfiguration.widgets.length) {
        return state;
      }

      const widgetsToUpdate = getWidgetsFromIds({
        widgetIds,
        widgets: state.dashboardConfiguration.widgets,
      });

      const unselectedWidgets = xorBy(
        state.dashboardConfiguration.widgets,
        widgetsToUpdate,
        'id'
      );

      const bottomZIndex = minBy(unselectedWidgets, 'z')?.z ?? 0;
      const maxSelectedZ = maxBy(widgetsToUpdate, 'z')?.z ?? 0;

      const zOffset = bottomZIndex - 1 - maxSelectedZ;

      const translatedWidgets = state.dashboardConfiguration.widgets.map(
        (widget) => ({
          ...widget,
          z: widgetIds.includes(widget.id) ? widget.z + zOffset : widget.z,
        })
      );

      state.dashboardConfiguration.widgets = translatedWidgets;
    },

    resizeWidgets: (
      state,
      action: PayloadAction<{
        anchor: Anchor;
        widgets: DashboardWidget[];
        vector: Position;
        complete?: boolean;
      }>
    ) => {
      const { anchor, widgets, vector, complete } = action.payload;

      const selectedWidgetIds = widgets.map((w) => w.id);

      const selectionBox = getSelectionBox(widgets);

      if (!selectionBox) return state;

      const newSelectionBox = resizeSelectionBox({
        selectionBox,
        anchor,
        vector,
        grid: {
          height: state.dashboardConfiguration.displaySettings.numRows,
          width: state.dashboardConfiguration.displaySettings.numColumns,
        },
      });

      const resizer = (widget: DashboardWidget) =>
        transformWidget(
          widget,
          selectionBox,
          complete ? createRectangle(newSelectionBox) : newSelectionBox
        );

      const updateWidgets = (widgets: DashboardWidget[]) =>
        widgets.map((widget) => {
          if (!selectedWidgetIds.includes(widget.id)) return widget;
          return resizer(widget);
        });

      state.dashboardConfiguration.widgets = updateWidgets(
        state.dashboardConfiguration.widgets
      );
    },

    moveWidgets: (
      state,
      action: PayloadAction<{
        widgets: DashboardWidget[];
        vector: Position;
        complete?: boolean;
      }>
    ) => {
      const { vector, complete, widgets } = action.payload;
      const selectedWidgetIds = action.payload.widgets.map((w) => w.id);
      const selectionBox = getSelectionBox(widgets);
      if (!selectionBox) return state;

      const newSelectionBox = moveSelectionBox({
        selectionBox,
        vector,
        grid: {
          height: state.dashboardConfiguration.displaySettings.numRows,
          width: state.dashboardConfiguration.displaySettings.numColumns,
        },
      });

      const mover = (widget: DashboardWidget) =>
        transformWidget(
          widget,
          selectionBox,
          complete ? createRectangle(newSelectionBox) : newSelectionBox
        );

      const updateWidgets = (widgets: DashboardWidget[]) =>
        widgets.map((widget) => {
          if (!selectedWidgetIds.includes(widget.id)) return widget;
          return mover(widget);
        });

      state.dashboardConfiguration.widgets = updateWidgets(
        state.dashboardConfiguration.widgets
      );
    },

    disableCanvas: (state) => {
      state.isCanvasEnabled = false;
    },

    enableCanvas: (state) => {
      state.isCanvasEnabled = true;
    },

    replaceDashboardConfiguration: (
      state,
      {
        payload,
      }: PayloadAction<{ dashboardConfiguration: DashboardConfiguration }>
    ) => {
      state.dashboardConfiguration = payload.dashboardConfiguration;
    },
  },
});

function getWidgetsFromIds({
  widgets,
  widgetIds,
}: {
  widgets: DashboardWidget[];
  widgetIds: string[];
}): DashboardWidget[] {
  return compact(
    widgetIds.map((id) => widgets.find((widget) => widget.id === id))
  );
}

function gridFromDisplaySettings({
  numColumns,
  numRows,
}: Pick<DashboardDisplaySettings, 'numColumns' | 'numRows'>): Grid {
  return createGrid({ width: numColumns, height: numRows });
}
