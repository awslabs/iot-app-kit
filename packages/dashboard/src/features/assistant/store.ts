import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import uniqBy from 'lodash-es/uniqBy';
import { v4 as uuid } from 'uuid';
import type { AssistantStateTypes } from '#types';

export interface AssistantState {
  state: AssistantStateTypes;
  conversationId: string;
  isChatbotOpen: boolean;
  callerComponentId?: string;
  action?: string;
  actionId?: string;
  mode: 'on' | 'off';
  selectedQueries: {
    widgetId: string;
    widgetType: string;
    selectedProperties: number;
  }[];
}

export const initialState: AssistantState = {
  state: 'PASSIVE',
  conversationId: uuid(),
  isChatbotOpen: false,
  mode: 'off',
  selectedQueries: [],
};

export const {
  reducer,
  actions: {
    toggleAssistantMode,
    toggleChatbot,
    cleanAssistant,
    assistantSelectWidgets,
    assistantDeselectWidgets,
    assistantCleanWidgetsSelection,
  },
} = createSlice({
  name: 'assistant',
  initialState: initialState,
  reducers: {
    toggleChatbot: (
      state,
      {
        payload,
      }: PayloadAction<{
        open: boolean;
        callerComponentId: string;
        action?: 'summarize' | 'divedeep';
      }>
    ) => {
      state.isChatbotOpen = payload.open;
      state.callerComponentId = payload.callerComponentId;
      state.action = payload.action;
      state.actionId = uuid();
    },

    toggleAssistantMode: (
      state,
      { payload }: PayloadAction<{ mode: 'on' | 'off' }>
    ) => {
      state.mode = payload.mode;
    },

    cleanAssistant: (state) => {
      state.conversationId = uuid();
      state.callerComponentId = undefined;
      state.action = undefined;
      state.selectedQueries = [];
    },

    assistantSelectWidgets: (
      state,
      {
        payload,
      }: PayloadAction<{
        widgetId: string;
        widgetType: string;
        selectedProperties: number;
      }>
    ) => {
      state.selectedQueries = uniqBy(
        [...state.selectedQueries, payload],
        'widgetId'
      );
    },

    assistantDeselectWidgets: (
      state,
      { payload }: PayloadAction<{ widgetId: string }>
    ) => {
      state.selectedQueries = state.selectedQueries.filter(
        ({ widgetId }) => widgetId !== payload.widgetId
      );
    },

    assistantCleanWidgetsSelection: (state) => {
      state.selectedQueries = [];
    },
  },
});
