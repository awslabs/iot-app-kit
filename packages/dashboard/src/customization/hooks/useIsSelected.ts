import includes from 'lodash/includes';
import map from 'lodash/map';
import { useCallback } from 'react';
import type { DashboardWidget } from '~/types';
import { useSelectedWidgets } from '~/hooks/useSelectedWidgets';

/**
 * Helper hook that can be exposed to consumers making their own widget components
 *
 * used to determine if this widget is in the selection
 *
 */
export const useIsSelected = <T extends DashboardWidget>(widget: T) => {
  const selectedWidgets = useSelectedWidgets();

  const isSelected = useCallback(
    () =>
      includes(
        map(selectedWidgets, (sw) => sw.id),
        widget.id
      ),
    [selectedWidgets, widget]
  );

  return isSelected();
};
