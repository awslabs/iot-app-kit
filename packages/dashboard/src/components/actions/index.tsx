import React, { memo } from 'react';
import { useDispatch } from 'react-redux';

import { Button, SpaceBetween, FormField } from '@cloudscape-design/components';
import { onToggleReadOnly } from '~/store/actions';
import type { DashboardMessages } from '~/messages';
import type { DashboardState } from '~/store/state';
import { isEqual, pick } from 'lodash';
import { DashboardSave } from '~/types';

export type ActionsProps = {
  messageOverrides: DashboardMessages;
  grid: DashboardState['grid'];
  readOnly: boolean;
  dashboardConfiguration: DashboardState['dashboardConfiguration'];
  onSave?: DashboardSave;
};

const Actions: React.FC<ActionsProps> = ({ dashboardConfiguration, messageOverrides, readOnly, onSave }) => {
  const dispatch = useDispatch();

  const handleOnSave = () => {
    if (!onSave) return;
    onSave(dashboardConfiguration);
  };

  const handleOnReadOnly = () => {
    dispatch(onToggleReadOnly());
  };

  return (
    <FormField label={messageOverrides.toolbar.actions.title}>
      <SpaceBetween size='s' direction='horizontal'>
        {onSave && <Button onClick={handleOnSave}>{messageOverrides.toolbar.actions.save}</Button>}
        <Button onClick={handleOnReadOnly}>{readOnly ? 'Edit' : 'Preview'}</Button>
      </SpaceBetween>
    </FormField>
  );
};

const gridAsComparable = (grid: DashboardState['grid']) => pick(grid, ['height', 'width', 'cellSize', 'stretchToFit']);
const actionsComparator = (a: Readonly<ActionsProps>, b: Readonly<ActionsProps>): boolean => {
  const gridIsSame = isEqual(gridAsComparable(a.grid), gridAsComparable(b.grid));
  const dashboardConfigurationIsSame = isEqual(a.dashboardConfiguration, b.dashboardConfiguration);
  return gridIsSame && dashboardConfigurationIsSame;
};

export default memo(Actions, actionsComparator);
