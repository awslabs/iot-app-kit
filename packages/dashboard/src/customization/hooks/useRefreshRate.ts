import { useDispatch, useSelector } from 'react-redux';
import { RefreshRate } from '~/components/querySettingsSync/types';
import { onUpdateRefreshRateAction } from '~/store/actions';
import { DashboardState } from '~/store/state';

export const useRefreshRate = () => {
  const dispatch = useDispatch();

  const updateRefreshRate = (refreshRate: RefreshRate) => {
    dispatch(
      onUpdateRefreshRateAction({
        refreshRate,
      })
    );
  };

  const refreshRate = useSelector(
    (state: DashboardState) =>
      state.dashboardConfiguration.querySettings?.refreshRate
  );

  return [refreshRate, updateRefreshRate] as const;
};
