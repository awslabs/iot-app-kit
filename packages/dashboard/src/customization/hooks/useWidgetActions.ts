import { useDispatch } from 'react-redux';

import { onUpdateWidgetsAction } from '~/store/actions';
import type { DashboardWidget } from '~/types';

/**
 * Helper hook that can be exposed to consumers making their own widget components
 *
 * used to update itself in the store
 *
 */
export const useWidgetActions = <T extends DashboardWidget>() => {
  const dispatch = useDispatch();

  const update = (widget: T) =>
    dispatch(
      onUpdateWidgetsAction({
        widgets: [widget],
      })
    );

  return {
    update,
  };
};
