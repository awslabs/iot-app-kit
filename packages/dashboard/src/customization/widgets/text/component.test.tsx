import React from 'react';

import { render } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import TextWidgetComponent from './component';
import { useIsSelected } from '~/customization/hooks/useIsSelected';
import { onChangeDashboardGridEnabledAction } from '~/store/actions';

jest.mock('~/store/actions', () => ({
  ...jest.requireActual('~/store/actions'),
  onChangeDashboardGridEnabledAction: jest.fn(),
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('~/customization/hooks/useIsSelected', () => ({
  useIsSelected: jest.fn(),
}));

jest.mock('./link', () => (props: unknown) => <div data-mocked='TextLink'>{JSON.stringify(props)}</div>);
jest.mock('./link/editableLink', () => (props: unknown) => (
  <div data-mocked='EditableTextLink'>{JSON.stringify(props)}</div>
));
jest.mock('./styledText/textArea', () => (props: unknown) => (
  <div data-mocked='StyledTextArea'>{JSON.stringify(props)}</div>
));
jest.mock('./styledText/editableText', () => (props: unknown) => (
  <div data-mocked='EditableStyledText'>{JSON.stringify(props)}</div>
));
jest.mock('./styledText', () => (props: unknown) => <div data-mocked='StyledText'>{JSON.stringify(props)}</div>);

describe('Text Widget', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  [
    { readOnly: true, isSelected: true, isUrl: true },
    { readOnly: true, isSelected: true, isUrl: false },
    { readOnly: true, isSelected: false, isUrl: false },
    { readOnly: false, isSelected: false, isUrl: false },
    { readOnly: false, isSelected: false, isUrl: true },
    { readOnly: false, isSelected: true, isUrl: true },
    { readOnly: false, isSelected: true, isUrl: false },
    { readOnly: true, isSelected: false, isUrl: true },
  ].forEach((configuration) => {
    it(`should render ${JSON.stringify(configuration)} correctly`, () => {
      const { isSelected, readOnly, isUrl } = configuration;
      (useIsSelected as jest.Mock).mockImplementation(() => isSelected);
      (useSelector as jest.Mock).mockImplementation(() => readOnly);
      (useDispatch as jest.Mock).mockReturnValue(jest.fn());

      const { container } = render(
        <TextWidgetComponent
          id='some-id'
          x={1}
          y={2}
          z={3}
          height={100}
          width={100}
          type='text-widget'
          properties={{ isUrl, value: 'abc' }}
        />
      );

      expect(container).toMatchSnapshot();
    });
  });

  it('should exit edit mode when unmounted', () => {
    (useDispatch as jest.Mock).mockImplementation(() => jest.fn((cb) => cb())); // short curcuit dispatch

    (useIsSelected as jest.Mock).mockImplementation(() => false);
    (useSelector as jest.Mock).mockImplementation(() => false);
    (useDispatch as jest.Mock).mockReturnValue(jest.fn());

    const { unmount } = render(
      <TextWidgetComponent
        id='some-id'
        x={1}
        y={2}
        z={3}
        height={100}
        width={100}
        type='text-widget'
        properties={{ isUrl: false, value: 'abc' }}
      />
    );
    unmount();

    expect(onChangeDashboardGridEnabledAction).toBeCalledWith({ enabled: true });
  });
});
