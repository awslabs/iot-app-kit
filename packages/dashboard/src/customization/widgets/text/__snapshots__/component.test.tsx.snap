// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Widget should render {"readOnly":false,"isSelected":false,"isUrl":false} correctly 1`] = `
<div>
  <div
    data-mocked="EditableStyledText"
  >
    {"readOnly":false,"isSelected":false,"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":false,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":false,"isSelected":false,"isUrl":true} correctly 1`] = `
<div>
  <div
    data-mocked="EditableTextLink"
  >
    {"readOnly":false,"isSelected":false,"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":true,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":false,"isSelected":true,"isUrl":false} correctly 1`] = `
<div>
  <div
    data-mocked="EditableStyledText"
  >
    {"readOnly":false,"isSelected":true,"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":false,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":false,"isSelected":true,"isUrl":true} correctly 1`] = `
<div>
  <div
    data-mocked="EditableTextLink"
  >
    {"readOnly":false,"isSelected":true,"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":true,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":true,"isSelected":false,"isUrl":false} correctly 1`] = `
<div>
  <div
    data-mocked="StyledText"
  >
    {"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":false,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":true,"isSelected":false,"isUrl":true} correctly 1`] = `
<div>
  <div
    data-mocked="TextLink"
  >
    {"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":true,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":true,"isSelected":true,"isUrl":false} correctly 1`] = `
<div>
  <div
    data-mocked="StyledText"
  >
    {"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":false,"value":"abc"}}
  </div>
</div>
`;

exports[`Text Widget should render {"readOnly":true,"isSelected":true,"isUrl":true} correctly 1`] = `
<div>
  <div
    data-mocked="TextLink"
  >
    {"id":"some-id","x":1,"y":2,"z":3,"height":100,"width":100,"type":"text-widget","properties":{"isUrl":true,"value":"abc"}}
  </div>
</div>
`;
